src/Mongolid/ActiveRecord.php

```diff
@@ -0,0 +1,177 @@
+<?php
+namespace Mongolid;
+
+use Mongolid\DataMapper\DataMapper;
+use Mongolid\Container\Ioc;
+use Mongolid\Model\Attributes;
+use Mongolid\Model\Relations;
+
+/**
+ * The Mongolid\ActiveRecord base class will ensure to enable your entity to
+ * have methods to interact with the database. It means that 'save', 'insert',
+ * 'update', 'where', 'first' and 'all' are available within every instance.
+ * The Mongolid\Schema that describes the entity will be generated on the go
+ * based on the $fields.
+ *
+ * @package  Mongolid
+ */
+abstract class ActiveRecord
+{
+    use Attributes, Relations;
+
+    /**
+     * Name of the collection where this kind of Entity is going to be saved or
+     * retrieved from
+     * @var string
+     */
+    public $collection = 'mongolid';
+
+    /**
+     * Describes the Schema fields of the model. Optionally you can set it to
+     * the name of a Schema class to be used.
+     * @see  Mongolid\Schema::$fields
+     * @var  string|string[]
+     */
+    protected $fields  = [
+        '_id' => 'objectId'
+    ];
+
+    /**
+     * The $dynamic property tells if the object will accept additional fields
+     * that are not specified in the $fields property. This is usefull if you
+     * doesn't have a strict document format or if you want to take full
+     * advantage of the "schemaless" nature of MongoDB.
+     * @var boolean
+     */
+    public $dynamic = true;
+
+    /**
+     * Saves this object into database
+     *
+     * @return boolean Success
+     */
+    public function save()
+    {
+        return $this->getDataMapper()->save($this);
+    }
+
+    /**
+     * Insert this object into database
+     *
+     * @return boolean Success
+     */
+    public function insert()
+    {
+        return $this->getDataMapper()->insert($this);
+    }
+
+    /**
+     * Updates this object in database
+     *
+     * @return boolean Success
+     */
+    public function update()
+    {
+        return $this->getDataMapper()->update($this);
+    }
+
+    /**
+     * Deletes this object in database
+     *
+     * @return boolean Success
+     */
+    public function delete()
+    {
+        return $this->getDataMapper()->delete($this);
+    }
+
+    /**
+     * Gets a cursor of this kind of entities that matches the query from the
+     * database
+     *
+     * @param  array  $query
+     *
+     * @return \Mongolid\Cursor\Cursor
+     */
+    public static function where(array $query = [])
+    {
+        return Ioc::make(get_called_class())
```

Comment:
Nice observation. But the `Model::where([])->first()` will not resolve twice, because the return of `where` and `all` will be a `Mongolid\Cursor`. So I'm not seeing much advantage of abstracting to an `resolveCalledClass()`